### PR TITLE
SONARJAVA-3687: S5979 should not report on classes annotated with JUnit5's @Nested

### PR DESCRIPTION
**Pull request by dorian-burihabwa-sonarsource**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/tests/MockitoAnnotatedObjectsShouldBeInitializedCheck.java [hunk 1]
```diff
@@ -73,16 +79,32 @@
   @Override
   public void visitNode(Tree tree) {
     ClassTree testClass = (ClassTree) tree;
+    if (visited.contains(testClass)) {
+      return;
+    }
+
+    if (isClassProperlyAnnotated(testClass)) {
+      List<ClassTree> children = getChildrenCoveredByAnnotations(testClass);
+      visited.addAll(children);
```
Comment (quentin-jaquier-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/tests/MockitoAnnotatedObjectsShouldBeInitializedCheck.java:88]: You can probably return here, and remove the redundant `isClassProperlyAnnotated` computation in `mocksAreProperlyInitialized`.

**Review by quentin-jaquier-sonarsource — CHANGES_REQUESTED** [review]
I'm wondering if the initialization via a Rule or during the setup in a parent works for the Nested class aswell. Did you check that?

File: java-checks/src/main/java/org/sonar/java/checks/tests/MockitoAnnotatedObjectsShouldBeInitializedCheck.java [hunk 2]
```diff
@@ -65,6 +69,8 @@
 
   private static final String MESSAGE = "Initialize mocks before using them.";
 
+  private Set<ClassTree> visited = new HashSet<>();
```
[the remark this answers is not in the document]
Comment (dorian-burihabwa-sonarsource) [reply java-checks/src/main/java/org/sonar/java/checks/tests/MockitoAnnotatedObjectsShouldBeInitializedCheck.java:72]: Renamed

File: java-checks/src/main/java/org/sonar/java/checks/tests/MockitoAnnotatedObjectsShouldBeInitializedCheck.java (hunk 1 above)
Comment (dorian-burihabwa-sonarsource) [reply java-checks/src/main/java/org/sonar/java/checks/tests/MockitoAnnotatedObjectsShouldBeInitializedCheck.java:88]: Modified

**Comment by dorian-burihabwa-sonarsource** [comment]
> I'm wondering if the initialization via a Rule or during the setup in a parent works for the Nested class aswell. Did you check that?

Experimenting with it, and looking up the different constructs, we can discard the @Rule solution as it's a JUnit4 construct and is not expected to be mixed with JUnit5's @Nested. It makes sense to cover `ExtendWith` as it comes from JUnit5. But `openMocks` and `initMocks` used in setup methods cannot cover nested classes as they need to target and instance of the class they run in (eg `MockitoAnnotations.openMocks(this)` or `MockitoAnnotations.openMocks(this.myMockedField)`) and thus cannot cover mocks in inner classes.

**Review by quentin-jaquier-sonarsource — APPROVED** [review]
LGTM.

Thanks for investigating what is or not possible.

**Pull request merged**